metadata/pkg/drivers/aws/aws.go

```diff
@@ -99,43 +98,46 @@ func ObjectList(sess *session.Session, bucket *model.MetaBucket) {
 	bucket.NumberOfObjects = numObjects
 	bucket.TotalSize = totSize
 	bucket.Objects = objectArray
+	return nil
 }
 
-func GetBucketMeta(buckIdx int, bucket *s3.Bucket, sess *session.Session, bucketArray []*model.MetaBucket, wg *sync.WaitGroup) error {
+func GetBucketMeta(buckIdx int, bucket *s3.Bucket, sess *session.Session, bucketArray []*model.MetaBucket, numBuckets *int, wg *sync.WaitGroup) error {
 	defer wg.Done()
-	buck := &model.MetaBucket{}
-	bucketArray[buckIdx] = buck
-	buck.CreationDate = *bucket.CreationDate
-	buck.Name = *bucket.Name
+
 	svc := s3.New(sess)
 	loc, err := svc.GetBucketLocation(&s3.GetBucketLocationInput{Bucket: bucket.Name})
 	if err != nil {
 		log.Errorf("unable to get bucket location. failed with error: %v", err)
 		return err
 	}
 
+	if *loc.LocationConstraint != *sess.Config.Region {
+		return nil
+	}
+
+	buck := &model.MetaBucket{}
+	buck.Name = *bucket.Name
+	buck.CreationDate = *bucket.CreationDate
 	buck.Region = *loc.LocationConstraint
 
-	if *sess.Config.Region != buck.Region {
-		svc = s3.New(sess, aws.NewConfig().WithRegion(buck.Region))
+	err = ObjectList(sess, buck)
+	if err != nil {
+		return err
 	}
 
 	tags, err := svc.GetBucketTagging(&s3.GetBucketTaggingInput{Bucket: bucket.Name})
 
 	if err != nil && !strings.Contains(err.Error(), "NoSuchTagSet") {
 		log.Errorf("unable to get bucket tags. failed with error: %v", err)
-	} else {
+	} else if err == nil {
 		tagset := make(map[string]string)
 		for _, tag := range tags.TagSet {
 			tagset[*tag.Key] = *tag.Value
 		}
 		buck.BucketTags = tagset
 	}
-
-	ObjectList(sess, buck)
-	if err != nil {
-		return err
-	}
+	bucketArray[buckIdx] = buck
+	*numBuckets += 1
```

Comment:
may cause race condition

Reply:
yes thanks for pointing out. its resolved now.